why-polars-for-data-cleaning/index.qmd

```diff
@@ -0,0 +1,185 @@
+---
+
+::: content-hidden
+Use other decision posts as inspiration to writing these. Leave the
+content-hidden sections in the text for future reference.
+:::
+
+## Context and problem statement
+
+::: content-hidden
+State the context and some background on the issue, then write a
+statement in the form of a question for the problem.
+:::
+
+When working with data analysis, data is rarely ready to be used 'as
+is'. There will always be things that are either very wrong (like text
+in what should be a number field), or slightly wrong (like a single
+missing value, or an outlier, which may or may not be a data entry
+error). In Sprout we are not worried about the single missing values or
+the outliers, but we are concerned with the data that contradicts the
+definition of the resource (eg. TBC in a date field). There are a lot of
+tools that can be used to both analyse a data set, and to clean it,
+either as part of the analysis, or as part of the initial check before
+the data is transformed and stored. The checking of the data prior to
+inclusion into Sprout is done within Sprout itself, this work will
+result in a list of errors which the user will need to handle before
+they can load the data into Sprout.
+
+> Which tool should we use for cleaning and processing data before
+> loading it into Sprout?
+
+## Decision drivers
+
+::: content-hidden
+List some reasons for why we need to make this decision and what things
+have arisen that impact work.
+:::
+
+-   We need a tool that can clean data and is easy to integrate with our
+    current workflow.
+-   We want to use an established tool that is well supported.
+-   We want a tool which users will find easy to learn and use.
+-   We do not want to build a tool from scratch.
+
+## Considered options
+
+::: content-hidden
+List and describe some of the options, as well as some of the benefits
+and drawbacks for each option.
+:::
+
+There are a number of tools available to use for data cleaning, but as
+we want something which will run in python, and which is well supported,
+we have narrowed our options down to the following:
+
+-   [DuckDB](https://www.duckdb.org/)
+-   [Pandas](https://pandas.pydata.org/)
+-   [Polars](https://pola.rs/)
+
+### DuckDB
+
+DuckDB is as the name implies not a Python library, but an in-process
+OLAP database solution. It is a very young tool, version 1.0.0 was
+released in 2024, although the first version was released in 2019. The
+developers have set it up as a not-for-profit foundation in the
+Netherlands.
+
+::: columns
+::: column
+#### Benefits
+
+-   Although a very young project it is already in use by a number of
+    large companies and has a large community of users.
+    -   The code is open source, and the way it has been set up is
+        promising for it to remain that way.
+-   It is very fast, and can handle large data sets.
+:::
+
+::: column
+#### Drawbacks
+
+-   It is not designed to be used as a data cleaning tool, but as a
+    database.
```

Comment:
I don't understand this comment. SQL is a language that does lots of data cleaning.

Reply:
It's more that it isn't the main focus of the tool, data cleaning is of course something sql is designed to also do, but the main purpose of a database is to take data in and store it.

Reply:
DuckDB is an in memory SQL tool, it doesn't store it in a database on the hard drive. Only acts like it does in memory 😊

Reply:
Or rather, it is used for it's in memory abilities. It can of course write to a file, and read from it.